client/src/components/SpotifyWidget/index.js

```diff
@@ -0,0 +1,196 @@
+import React, { useState, useEffect } from "react";
+import { makeStyles } from "@material-ui/core";
+import Button from "@material-ui/core/Button";
+import IconButton from "@material-ui/core/IconButton";
+import style from "./style.module.scss";
+import PlayArrowIcon from "@material-ui/icons/PlayArrow";
+import FastForwardIcon from "@material-ui/icons/FastForward";
+import FastRewindIcon from "@material-ui/icons/FastRewind";
+import PauseIcon from "@material-ui/icons/Pause";
+import { accessUrl, getAuthToken, getRefreshedToken, setExpiryTime } from "./spotify";
+import SpotifyWebApi from "spotify-web-api-js";
+
+const spotify = new SpotifyWebApi();
+
+const useStyles = makeStyles({
+    root: {
+        margin: "30px 20px",
+        background: "#1DB954",
+        borderRadius: 50,
+        padding: "4px 25px",
+        width: "calc(100% - 40px)",
+    },
+    label: {
+        textTransform: "capitalize",
+        fontSize: "24px",
+    },
+});
+
+export default function SpotifyWidget() {
+    const classes = useStyles();
+    const [play, setPlay] = useState(false);
+    const [isLoggedIn, setIsLoggedIn] = useState(false);
+    const [item, setItem] = useState(false);
+
+    useEffect(() => {
+        var _token = localStorage.getItem("spotify-token");
+
+        if (!_token) {
+            setIsLoggedIn(false);
+
+            // the authorisation code comes from the query of page's url after the user logs in to spotify
+            // check to see if the code is there. If it is, then use it get the access token
+            if (window.location.search !== "") {
+                getAuthToken().then((tokenData) => {
+                    // store the tokens and expiry time in local storage
+                    _token = tokenData.access_token;
+                    const refreshToken = tokenData.refresh_token;
+
+                    localStorage.setItem("spotify-token", _token);
+                    localStorage.setItem("spotify-refresh-token", refreshToken);
+                    setExpiryTime(tokenData.expires_in);
+
+                    // the authorisation code is no longer needed, remove it from the URl
+                    window.location.search = "";
+                });
+            }
+        }
+
+        if (_token) {
+            var token_expiration = localStorage.getItem("spotify-token-expiration");
+            localStorage.setItem("spotify-token-expiration", Date.now());
```

Comment:
Why is this being done? It will mean that the next time the component re-mounts, it will think the token has expired, even if the original expiry time hasn't passed yet.

Reply:
oops, thats from my testing it shouldn't be there